### PR TITLE
Add UI Page for OIDC Configuration

### LINKED ISSUE #2: Admin page
Tracking issue for the admin page

### PR DESCRIPTION
**Pull request by matthewstrasiotto**
Partially resolves #2  

For now, this only implements OID configuration, though gives a solid reference for SAML configuration, which will be similar. 

Presently, options that take json objects require json to be given on the UI - For list items, this would be better as a multiline text field that splits on newlines, and more more complex mappings, something actually good needs to be made - but this will do for now.

### CODE REVIEW

**Comment by 9p4** [comment]
Could you make the CI happy (it's all style)?

**Comment by matthewstrasiotto** [comment]
> Could you make the CI happy (it's all style)?

Yep, I've updated my own build command to use --warnaserror too

**Comment by matthewstrasiotto** [comment]
Looks like it should pass CI checks now, seems to on my fork: 

https://github.com/matthewstrasiotto/jellyfin-plugin-sso/pull/2

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [inline SSO-Auth/Config/configPage.html:42]: I've been getting a TypeErro on these, I think I need to add a label attribute to all of the `is="emby-input"` fields

**Comment by matthewstrasiotto** [comment]
Actually, please don't merge this just yet, until https://github.com/9p4/jellyfin-plugin-sso/pull/18/files#r857104729 is resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:42]: Just returned to this, it's actually not because of a missing label attribute, its because of this https://github.com/jellyfin/jellyfin-web/issues/3592

Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:42]: Removing the `class="emby-input"` should fix it

**Comment by matthewstrasiotto** [comment]
Should be RTM - The frontend issues are resolved now

File: SSO-Auth/Config/config.js
Comment (9p4) [inline SSO-Auth/Config/config.js:5]: The CI doesn't catch this, but there seems to be a lot of extra whitespace and newlines where there shouldn't be. Could you fix that?

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:4]: Lots of extra (unneeded) newlines here too

**Review by 9p4 — CHANGES_REQUESTED** [review]

File: SSO-Auth/Config/config.js
Comment (matthewstrasiotto) [reply SSO-Auth/Config/config.js:5]: Yeah, the CI doesn't lint the js. 

I'll run a linter against the js and html, stand by for that (no need to make any more style comments until I address this)

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:4]: Ran a formatter against this, as well as the js

File: SSO-Auth/Config/config.js
Comment (matthewstrasiotto) [reply SSO-Auth/Config/config.js]: Done

**Review by 9p4 — APPROVED** [review]

**Comment by matthewstrasiotto** [comment]
This'll probably be annoying to rebase but that's what you get for not formatting your code, I've thrown both the html & js at prettier so far

**Comment by 9p4** [comment]
I'm going to take this for a test run and merge once it all feels good :)

Thanks for the PR!

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:15]: Can the contrast of this be improved as well as the padding be fixed?

**Comment by matthewstrasiotto** [comment]
I'll wait to rebase #19 until I wait & see if you have more changes to request. 

Happy to say that I've successfully used this to configure my dev server to authorize against my authelia instance, and was able to login with the most basic settings:

```yaml
---
authelia:
  OidEndpoint: https://authelia.example.com/.well-known/openid-configuration/
  OidClientId: jellyfin
  OidSecret: <omitted>
  Enabled: true
  EnableAuthorization: true
  EnableAllFolders: true
  EnabledFolders: []
  Roles: []
  EnableFolderRoles: false
  FolderRoleMapping: []
  RoleClaim: a.b.d
```

File: SSO-Auth/Config/configPage.html
Comment (9p4) [reply SSO-Auth/Config/configPage.html:15]: The string probably should also be changed (no longer just configured via API)

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:23]: Should not be a sentence and instead a phrase (if it's ending w a colon)

Maybe something like "OIDC Provider:"

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:46]: Can there also be an "Add Provider" button?

File: SSO-Auth/Config/configPage.html
Comment (9p4) [reply SSO-Auth/Config/configPage.html:74]: Better yet, the help string here isn't needed (redundant with the title string)

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:65]: Should be "Name of OpenID Provider:"

**Comment by matthewstrasiotto** [comment]
So far I've noticed that there's not as much validation of configuration as there probably should be, particularly the json fields, whereby if they are left empty, there'll be runtime errors when you try to run your flow. 

For now, the json fields that you want to leave unset seem to need some parse-able placeholder, `[]` or `{}`

File: SSO-Auth/Config/configPage.html
Comment (9p4) [reply SSO-Auth/Config/configPage.html:15]: Should become a help or advanced config button

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:94]: Capitalization: "OpenID Client ID:"

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:102]: Redundant (remove)

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:110]: Shouldn't be capitalized like this if it's a sentence. "The OpenID secret. Randomly generated and shared."

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:120]: Redundant

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:131]: This may be a little vague, but the description helps. Let's think about changing this string too.

**Comment by matthewstrasiotto** [comment]
Also, future work to make this actually feel nice is going to involve:

- client (or server) side mapping of folders to their respective ids
- a nice(er) UI for selecting folders, using LDAP auth as a reference:
image

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:166]: Instead of asking for JSON, maybe use what the LDAP plugin uses? https://github.com/jellyfin/jellyfin-plugin-ldapauth/blob/476156b8fae962e67daa92079e4e0c5273335455/LDAP-Auth/Config/configPage.html#L132

File: SSO-Auth/Config/configPage.html
Comment (9p4) [reply SSO-Auth/Config/configPage.html:191]: Should also not be taking raw JSON but instead something more user-friendly.

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:192]: > priveleges

Spelling

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:203]: Consistency. Should be something like "Enable Role-Based Folder Access:"

File: SSO-Auth/Config/configPage.html
Comment (9p4) [inline SSO-Auth/Config/configPage.html:217]: Should not take raw JSON and instead be more user-friendly.

**Review by 9p4 — CHANGES_REQUESTED** [review]
A lot of nitpicky UX changes. Most are spelling/grammar/consistency but a few are for how data is input and may take a while to change.

**Comment by matthewstrasiotto** [comment]
@94p although I'd like to get the input changes in (json input is def not user friendly), I'd still call this an improvement over the previous system (no config page). 
Will you consider pulling this as good-enough-for-now if I get the spelling / punctuation / help string comments fixed?

**Comment by 9p4** [comment]
Yep!

**Comment by 9p4** [comment]
I would prefer for an "Add" button to be added before I merge, however.

**Comment by matthewstrasiotto** [comment]
> I would prefer for an "Add" button to be added before I merge, however.

If you mean to add a new oidc config, that's implemented- you fill out the form and press "add", at the bottom of the form.
If there's an entry loaded up, it updates that entry, but really it just writes to whatever entry has the corresponding "provider name"

I think I know what you're saying, though - to make the pattern more obvious, I could add a button up top "add" that just clears the form and maybe puts "new provider" into the name field, so the flow to add a new provider feels more obvious

**Comment by matthewstrasiotto** [comment]
When I get a second tomorrow, I'll address the style and spelling nits and then it'll be good to merge, with a clear path for how to get it usable enough to someday be beta-ready (I'll probably implement SAML config around the same time)

**Comment by 9p4** [comment]
Again, thank you so much for this contribution!

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:131]: I think taking the help text from the LDAP plugin will help- same concept, but they've expressed it clearer

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:15]: > Can the contrast of this be improved as well as the padding be fixed?

If you mean the link to readme, yes, the contrast is bad - I think I'll change it to a help button, and add a disclaimer that this is in alpha, doesn't have SAML, and to check the help for more info. 

As for padding, I'll add a proper title, and between that and the disclaimer it should look more polished

LDAP for reference:
```html
                    <div class="verticalSection verticalSection-extrabottompadding">
                        <div class="sectionTitleContainer flex align-items-center">
                            <h2 class="sectionTitle">LDAP Settings:</h2>
                            <a is="emby-button" class="raised button-alt headerHelpButton" target="_blank" href="https://github.com/jellyfin/jellyfin-plugin-ldapauth">${Help}</a>
                        </div>
                        <p><i>Note:</i> Making changes to this configuration requires a restart of Jellyfin.</p>
```

image

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:191]: A good reminder to check my privilege

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:74]: Actually, I think I can make the help string more useful - I'll clarify that this is how Jellyfin identifies the provider (As opposed to say, client id) and maybe that pressing save will create the provider if it exists, or update its settings if it doesnt't.

**Comment by matthewstrasiotto** [comment]
Okay, awake now, I'll address those things - Not fully clear on the "Add" functionality I described here

> I think I know what you're saying, though - to make the pattern more obvious, I could add a button up top "add" that just clears the form and maybe puts "new provider" into the name field, so the flow to add a new provider feels more obvious

I think if I improve the labeling of the add / update form, then it shouldn't be as necessary to implement a specific "Add" button, it should be clear what form does what.

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html:131]: For now I won't touch it, actually

**Comment by matthewstrasiotto** [comment]
Current UI 

image

By default, "select existing" is collapsed.

image

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved (In the sense that will be improved in later iteration)

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Shelved for now (WIll be revisted)

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved - WIll revisit help text in further iterations

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by
https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfR117-R125 

```html
                  <div class="fieldDescription">
                    The name used by Jellyfin to identify the OID provider.
                    <br />
                    If an OID provider with a matching name does not exist, a
                    new provider with this name will be created.
                    <br />
                    If an OID provider with a matching name already exists, the
                    settings for that provider will be updated.
                  </div>
```

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Spelling comment resolved by https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/5112bb13afe9943aad8784e551d7277870ddeaae

Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Json comment to be revisited in future iterations.

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL59-R61 resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL59-R61 resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL148-R148 resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL156-R160 resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL169-R173 resolved

File: SSO-Auth/Config/configPage.html
Comment (matthewstrasiotto) [reply SSO-Auth/Config/configPage.html]: Resolved https://github.com/9p4/jellyfin-plugin-sso/pull/18/commits/3bc8cc9a85dd51bd1f18dfb597b16a99d6197a2f#diff-d1cae9c0a00116bf22523395cbfe6b7f5d9603ef680a7202c05e4a53bc3b05cfL180-L182

**Comment by matthewstrasiotto** [comment]
Documenting comments / requests that won't be actioned this PR in https://github.com/9p4/jellyfin-plugin-sso/issues/2#issuecomment-1101945157

**Review by 9p4 — APPROVED** [review]

**Comment by 9p4** [comment]
I'll do some more testing this week but looks very good!

**Pull request merged**
